Fix: ignore null literals for non nullable types

XinyiQiao opened the pull request:
Hi team, could you please review this draft PR on issue [#1701](https://github.com/questdb/questdb/issues/1701)?

We fixed this issue by always return:
false for col = null when col's type is [char, byte, bool, short]
true for col != null when col's type is [char, byte, bool, short]

For the test` io.questdb.griffin.InsertNullTest#testInsertNull`, we had an issue with \n in query result and our current fix was to remove all \n characters in TestUtils.assertEquals. Not sure if this follows the best practice and would appreciate any insight!

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/questdb/questdb?pullRequest=2000) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/questdb/questdb?pullRequest=2000) before we can accept your contribution.<br/>**0** out of **3** committers have signed the CLA.<br/><br/>:x: Haotian Zheng<br/>:x: XinyiQiao<br/>:x: haotiaz<br/><hr/>**Haotian Zheng** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/questdb/questdb?pullRequest=2000) it.</sub>

marregui commented:
Thank you very much for the contribution!. The approach is correct. 

Looking into the subject, it turns out that type `char` needs in fact to be nullable. At present it is not documented as such (I will update the documentation), however the argument is that a char is a one character string, and the latter is nullable. In both cases the empty value `''` and NULL should be the same. Type `char` is nullable. 

These tests would need to pass:

```
create table tab_char(v char);
insert into tab_char select rnd_char() from long_sequence(9);
insert into tab_char values(NULL);
tab_char where v is NULL --> 1 row ;
tab_char where v is NOT NULL --> 9 rows ;
select cast(0 AS CHAR) IS NULL --> true;
select cast(0L AS CHAR) IS NULL --> true;
select cast('' AS CHAR) IS NULL --> true;
select cast(NULL AS CHAR) IS NULL --> true;
select cast(false AS CHAR) IS NULL --> false;
select cast(0 AS CHAR) IS NOT NULL --> false;
select cast(0L AS CHAR) IS NOT NULL --> false;
select cast('' AS CHAR) IS NOT NULL --> false;
select cast(NULL AS CHAR) IS NOT NULL --> false;
select cast(false AS CHAR) IS NOT NULL --> true;
```

You could have a look at the above, or simply revert/ignore `char` related code in this PR, to make it about the only non nullable types `short`, `byte` and `boolean`.

For `short` these tests would need to pass:

```
create table tab_short(v short);
insert into tab_short select rnd_short() from long_sequence(9);
insert into tab_short values(NULL);
tab_short where v is NULL --> empty ;
tab_short where v is NOT NULL --> all 10 rows ;
select cast(0 AS SHORT) IS NULL --> false;
select cast(0L AS SHORT) IS NULL --> false;
select cast('' AS SHORT) IS NULL --> false;
select cast(NULL AS SHORT) IS NULL --> false;
select cast(false AS SHORT) IS NULL --> false;
select cast(0 AS SHORT) IS NOT NULL --> true;
select cast(0L AS SHORT) IS NOT NULL --> true;
select cast('' AS SHORT) IS NOT NULL --> true;
select cast(NULL AS SHORT) IS NOT NULL --> true;
select cast(false AS SHORT) IS NOT NULL --> true;
```

For `byte` these tests would need to pass:

```
create table tab_byte(v byte);
insert into tab_byte select rnd_byte() from long_sequence(9);
insert into tab_byte values(NULL);
tab_byte where v is NULL --> empty ;
tab_byte where v is NOT NULL --> all 10 rows ;
select cast(0 AS BYTE) IS NULL --> false;
select cast(0L AS BYTE) IS NULL --> false;
select cast('' AS BYTE) IS NULL --> false;
select cast(NULL AS BYTE) IS NULL --> false;
select cast(false AS BYTE) IS NULL --> false;
select cast(0 AS BYTE) IS NOT NULL --> true;
select cast(0L AS BYTE) IS NOT NULL --> true;
select cast('' AS BYTE) IS NOT NULL --> true;
select cast(NULL AS BYTE) IS NOT NULL --> true;
select cast(false AS BYTE) IS NOT NULL --> true;
```

For `boolean` these tests would need to pass, however there are a couple that hit the same bug. You can annotate tests to be ignored with `@Ignore` (please also add a `TODO` explaining that it evidences a bug). Alternatively you can go down the rabbit hole and get a closer look at the bug, which may be simple to fix in this PR.

```
create table tab_boolean(v boolean);
insert into tab_boolean select rnd_boolean() from long_sequence(9);
insert into tab_boolean values(NULL);
tab_boolean where v is NULL --> empty ;
tab_boolean where v is NOT NULL --> all 10 rows ;
select cast(0 AS BOOLEAN) IS NULL --> false;
select cast(0L AS BOOLEAN) IS NULL --> false;
select cast('' AS BOOLEAN) IS NULL --> false;              <---- HERE
select cast(NULL AS BOOLEAN) IS NULL --> false;
select cast(false AS BOOLEAN) IS NULL --> false;
select cast(0 AS BOOLEAN) IS NOT NULL --> true;
select cast(0L AS BOOLEAN) IS NOT NULL --> true;
select cast('' AS BOOLEAN) IS NOT NULL --> true;       <---- HERE
select cast(NULL AS BOOLEAN) IS NOT NULL --> true;
select cast(false AS BOOLEAN) IS NOT NULL --> true;
```

Thank you again @XinyiQiao , would you be able to update the PR.

As per the CLA issue, hmm, would you be able to update your fork and then make these five commits a single commit authored by you only?, I suppose you would then need to do a push force.

XinyiQiao commented:
Hi Miguel, thank you so much for your feedback! I've reverted char related code and working on the tests now!

XinyiQiao commented:
Hi Miguel @marregui I've added the tests you mentioned and reverted the change in TestUtils. Could you please have a look at the updated PR? Thank you!

marregui commented on core/src/main/java/io/questdb/griffin/engine/functions/eq/EqBooleanFunctionFactory.java:
```diff
@@ -61,6 +61,9 @@ public Func(Function left, Function right) {
         @Override
         public boolean getBool(Record rec) {
+            if(left.isNullConstant() || right.isNullConstant()) {
+                return negated || false;
```
Table of Truth:

| negated | result |
|---------|-------|
| false      | false   |
| true       | true     |

the `|| false` is thus redundant.

marregui commented on core/src/main/java/io/questdb/griffin/engine/functions/eq/EqByteFunctionFactory.java:
```diff
@@ -61,6 +61,9 @@ public Func(Function left, Function right) {
         @Override
         public boolean getBool(Record rec) {
+            if(left.isNullConstant() || right.isNullConstant()) {
+                return negated || false;
```
ditto 

Table of Truth:

| negated | result |
|---------|-------|
| false      | false   |
| true       | true     |

the `|| false` is thus redundant.

marregui commented on core/src/main/java/io/questdb/griffin/engine/functions/eq/EqShortFunctionFactory.java:
```diff
@@ -61,6 +61,9 @@ public Func(Function left, Function right) {
         @Override
         public boolean getBool(Record rec) {
+            if (left.isNullConstant() || right.isNullConstant()) {
+                return negated || false;
```
ditto 

Table of Truth:

| negated | result |
|---------|-------|
| false      | false   |
| true       | true     |

the `|| false` is thus redundant.

marregui commented on core/src/test/java/io/questdb/griffin/NullLiteralsTest.java:
```diff
@@ -0,0 +1,303 @@
+ ******************************************************************************/
+
+package io.questdb.griffin;
+import org.junit.Test;
```
needs a blank line between package declaration and first import.

marregui commented on core/src/test/java/io/questdb/griffin/InsertNullTest.java:
```diff
@@ -234,4 +234,18 @@ static String expectedNullInserts(String header, String nullValue, int count) {
         return sb.toString();
     }
+
+    static String expectedNullInserts(String header, String type, String nullValue, int count, boolean eq) {
```
I would simply use one method called `expectedNullInserts` like so:

```
static String expectedNullInserts(String header, String nullValue, int count, boolean expectsOutput) {
        StringSink sb = Misc.getThreadLocalBuilder();
        sb.put(header);
        if (expectsOutput) {
            for (int i = 0; i < count; i++) {
                sb.put(nullValue).put("\n");
            }
        }
        return sb.toString();
    }

    private static boolean isNotNullable(String type) {
        return Chars.equalsLowerCaseAscii(type, "short") ||
                Chars.equalsLowerCaseAscii(type, "byte") ||
                Chars.equalsLowerCaseAscii(type, "boolean");
    }
```

then you can call it:

```
testInsertNullThenFilterIsNull: expectedNullInserts("value\n", type[1], NULL_INSERTS, !isNotNullable(type[0])),
testInsertNullThenFilterIsNotNull: expectedNullInserts("value\n", type[1], NULL_INSERTS, isNotNullable(type[0])),
```

Note, you will need to amend InsertNullGeoHashTest (add true at the end in L150)

marregui reviewed (changes requested):
Overall very good, we are almost at the point of merging, how exciting!!!!! thank you for this Isabella.

bluestreak01 commented:
Hi Isabella, thank you for the PR! I had a look and it appears that the `Eq...` functions are not implemented to be efficient. When either of the args is `null constant` it renders the entire function `constant`. QuestDB SQL optimiser finds such constant functions and removes them from per-row execution.

The current implementations will be executing `where short = null` for each row. This should not be the case (it is `false` for every row).

I would suggest to move check for `null constant` to the factories. These factories should return `BooleanConstant.FALSE` if either of the args is `null constant`

XinyiQiao commented:
Hi Vlad @bluestreak01 thank you so much for the feedback! Understood that the factories should return BooleanConstant.FALSE for short = null case. For the short != null case, I believe it should be true for every row. Could you advise on how to approach the != case? Where would be the best place to check for the negated flag?

XinyiQiao commented:
> Hi Isabella, thank you for the PR! I had a look and it appears that the `Eq...` functions are not implemented to be efficient. When either of the args is `null constant` it renders the entire function `constant`. QuestDB SQL optimiser finds such constant functions and removes them from per-row execution.
> 
> The current implementations will be executing `where short = null` for each row. This should not be the case (it is `false` for every row).
> 
> I would suggest to move check for `null constant` to the factories. These factories should return `BooleanConstant.FALSE` if either of the args is `null constant`

Hi Vlad @bluestreak01 thank you so much for the feedback! I've moved the check for `null constant` to the factories. Please let me know if any additional issues need to be addressed.

marregui reviewed (approved):
LGTM well done @XinyiQiao

marregui commented:
Hi @XinyiQiao , could you please update your fork and the branch, thank you in advance.

bluestreak01 reviewed (changes requested):
performance improvement is required
